emma-language/src/main/scala/org/emmalanguage/ast/AST.scala

```diff
@@ -55,38 +55,56 @@ trait AST extends CommonAST
   /** Virtual non-overlapping semantic AST based on Scala trees. */
   object api extends API
 
+  /** Class for tree transformation definition. Additional field for name definition. */
+  case class TreeTransform(name: String = "", transformation: (u.Tree => u.Tree)) {
```

Comment:
> and then you don't need to override andThen and compose, unless we need to refine the return type of those to TreeTransform.

I think we do need to refine, if we would like to have a new name from the `andThen`. E.g., in the current version the new name of `a andThen b` is `a.name + " <<then>> " + b.name`

Reply:
I am not sure whether this is needed. The idea of the extension was to make top-level transformations (i.e., the ones used in the `pipeline` methods in `FlinkCompiler` and `SparkCompiler` named). This should be doable without a smaller diff.

Reply:
I guess most of those transformations that are chained together with other transforms using `andThen` are also used directly in a pipeline somewhere else (e.g., in tests). For example, see the usages of `Comprehension.resugarDataBag`.

In these cases we have to make them a `TreeTransform`, because of the direct usages in a pipeline call. But then, if they are already a `TreeTransform`, then why not let the names also propagate through `andThen`s, with just this simple overload?

Reply:
I am trying to keep some perspective of the original motivation.

The PR makes sense as a first step towards extending the macros so they write out the time for each step of the pipeline in a log.

To do this, we don't need extensive changes.

- introduce a class `TreeTransform` class in `Compiler`, abstracting over named `Tree => Tree` functions;
- introduce an implicit conversion that lifts pure `Tree => Tree` functions as `TreeTransform` with a default name like "anonymous" in the companion object;
- change the type of the `pipeline` argument from `Seq[Tree => Tree] to `Seq[Transform]`;
- wrap methods like `Core.anf` in a `TreeTransform` constructor, if possible without changing the formatting and indentation of the definition itself.

I really don't see a point of adding more code the one doing the 4 things above, unless there is a clear need to do so.

Reply:
I fully agree with you about fixing the indentation.

With the `andThen` stuff, I feel like we are debating something like <20 lines of code (the `andThen` itself, and putting the `TreeTransfor` call on ~10-15 transforms), which is basically already done by Lorand anyway.

I think the PR looks a bit scary now only because of the indentation mess. So I would suggest that we look at the PR again after the indentation is fixed, and then we will see everything more clearly.

Reply:
Agreed. Let's see the cleaner diff after the rework and see whether `andThen` and `compose` are needed.

Reply:
I cleaned up the line changes. Reduced it to pretty much two lines per method.

Looking into the other suggestions right now.